core-api/src/main/java/com/optimizely/ab/event/BatchEventProcessor.java

```diff
@@ -0,0 +1,204 @@
+ *
+ * Events passed to the BatchEventProcessor are immediately added to a BlockingQueue.
+ *
+ * The BatchEventProcessor maintains a single consumer thread that pulls events off of
+ * the BlockingQueue and buffers them for either a configured batch size or for a
+ * maximum duration before they resulting LogEvent is sent to the NotificationManager.
+ */
+public class BatchEventProcessor implements EventProcessor, AutoCloseable {
+
+    private static final Logger logger = LoggerFactory.getLogger(BatchEventProcessor.class);
+
+    public static final String CONFIG_BATCH_SIZE     = "event.processor.batch.size";
+    public static final String CONFIG_BATCH_DURATION = "event.processor.batch.duration.ms";
+
+    public static final int  DEFAULT_BATCH_SIZE     = 50;
+    public static final long DEFAULT_BATCH_DURATION = TimeUnit.MINUTES.toMillis(1);
+
+    private static final Object SHUTDOWN_SIGNAL = new Object();
+
+    private final NotificationManager<LogEvent> notificationManager = new NotificationManager<>();
+    private final BlockingQueue<Object> eventQueue;
+
+    private final int batchSize;
+    private final long batchDuration;
+    private final ExecutorService executor;
+
+    private Future<?> future;
+    private boolean isStarted = false;
+
+    // TODO use a builder.
+    public BatchEventProcessor() {
+        this(new ArrayBlockingQueue<>(1000));
+    }
+
+    public BatchEventProcessor(BlockingQueue<Object> eventQueue) {
+        this(eventQueue, null, null, Executors.newSingleThreadExecutor());
+    }
+
+    public BatchEventProcessor(BlockingQueue<Object> eventQueue, Integer batchSize, Integer batchDuration, ExecutorService executor) {
+        this.eventQueue = eventQueue;
+        this.batchSize = batchSize == null ? PropertyUtils.getInteger(CONFIG_BATCH_SIZE, DEFAULT_BATCH_SIZE) : batchSize;
+        this.batchDuration = batchDuration == null ? PropertyUtils.getLong(CONFIG_BATCH_DURATION, DEFAULT_BATCH_DURATION) : batchDuration;
+
+        if (executor == null) {
+            final ThreadFactory threadFactory = Executors.defaultThreadFactory();
+            this.executor = Executors.newSingleThreadExecutor(runnable -> {
+                Thread thread = threadFactory.newThread(runnable);
+                thread.setDaemon(true);
+                return thread;
+            });
+        } else {
+            this.executor = executor;
+        }
+
+        start();
+    }
+
+    public synchronized void start() {
+        if (isStarted) {
+            logger.info("Executor already started.");
+            return;
+        }
+
+        isStarted = true;
+        EventConsumer runnable = new EventConsumer();
+        future = executor.submit(runnable);
+    }
+
+    @Override
+    public void close() throws Exception {
+        logger.info("Start close");
+        eventQueue.put(SHUTDOWN_SIGNAL);
+        try {
+            future.get(5, TimeUnit.SECONDS);
```

Comment:
Will batched events get lost when this is called? Should we attempt to send batched events at this point?

Reply:
This is a blocking call waiting for the task to complete, i.e. process all pending events. The task will gracefully stop once it receives the `SHUTDOWN_SIGNAL` message, so in theory there will be no data loss. 

That being said we can't block indefinitely so the timeout is there as a maximum wait time to flush on close, at which point events will get dropped. I should make that configurable since 5 seconds might be too tight if the buffer has a lot of pending events.

Reply:
Added https://optimizely.atlassian.net/browse/OASIS-4975 to track.

Reply:
So we won't flush immediately, but instead just allow any scheduled flushes to occur up until the timeout? In JS, we flush immediately on close. Is there a reason not to flush immediately?

Reply:
The `SHUTDOWN_SIGNAL` will trigger a flush once consumed by the batching task. So this will be "immediate" if the queue is empty or it will be delayed until the consumer finishes draining the queue. I believe this is the proper behavior in a multi-threaded environment.